mapper/orm/src/main/java/org/hibernate/search/mapper/orm/automaticindexing/filter/impl/HibernateOrmAutomaticIndexingTypeFilterContext.java

```diff
@@ -0,0 +1,158 @@
+	}
+
+	@Override
+	public PojoAutomaticIndexingTypeFilterContext include(Class<?> clazz) {
+		addIfNotPresentInOther(
+				contextProvider.indexedWithSuperTypesByExactClass().getOrFail( clazz ),
+				includes,
+				excludes
+		);
+		return this;
+	}
+
+	@Override
+	public PojoAutomaticIndexingTypeFilterContext exclude(String name) {
+		addIfNotPresentInOther(
+				contextProvider.byEntityName().getOrFail( name ).typeIdentifier(),
+				excludes,
+				includes
+		);
+		return this;
+	}
+
+	@Override
+	public PojoAutomaticIndexingTypeFilterContext exclude(Class<?> clazz) {
+		addIfNotPresentInOther(
+				contextProvider.indexedWithSuperTypesByExactClass().getOrFail( clazz ),
+				excludes,
+				includes
+		);
+		return this;
+	}
+
+	public HibernateOrmAutomaticIndexingTypeFilter createFilter() {
+		return createFilter( null );
+	}
+
+	public HibernateOrmAutomaticIndexingTypeFilter createFilter(PojoAutomaticIndexingTypeFilterHolder fallback) {
+		Set<PojoRawTypeIdentifier<?>> allIncludes = new HashSet<>();
+		Set<PojoRawTypeIdentifier<?>> allExcludes = new HashSet<>();
+		boolean allTypesProcessed = true;
+
+		for ( PojoTypeContext<?> typeContext : contextProvider.byEntityName().values() ) {
+			PojoRawTypeIdentifier<?> typedIdentifier = typeContext.typeIdentifier();
+
+			//.include(...) => included no matter what; subclasses included unless excluded explicitly in the same filter
+			//.exclude(...) => excluded no matter what; subclasses excluded unless included explicitly in the same filter
+			//.include(...) + .exclude(...) for the same entity in the same filter => exception
+			// neither .include(...) nor .exclude(...) => defer to the superclass inclusions/exclusions,
+			// then to the application filter, with the same behavior, and by default include
+			if ( excludes.contains( typedIdentifier ) ) {
+				allExcludes.add( typedIdentifier );
+			}
+			else if ( includes.contains( typedIdentifier ) ) {
+				allIncludes.add( typedIdentifier );
+			}
+			else {
+				Optional<Class<?>> closestInclude = findClosestClass( typedIdentifier.javaClass(), includes );
+				Optional<Class<?>> closestExclude = findClosestClass( typedIdentifier.javaClass(), excludes );
+
+				if ( closestInclude.isPresent() && closestExclude.isPresent() ) {
+					// if include is a subclass of exclude - then include is more specific, and we allow indexing:
+					if ( closestExclude.get().isAssignableFrom( closestInclude.get() ) ) {
+						allIncludes.add( typedIdentifier );
+					}
+					else {
+						allExcludes.add( typedIdentifier );
+					}
+				}
+				else if ( closestExclude.isPresent() ) {
+					allExcludes.add( typedIdentifier );
+				}
+				else if ( closestInclude.isPresent() ) {
+					allIncludes.add( typedIdentifier );
+				}
+				else {
+					// if we don't find either include or exclude then we will defer the decision to either app filter or to a default (include)
+					// but that will happen in the filter.
+					allTypesProcessed = false;
+				}
+			}
```

Comment:
A slightly simpler (I think) algorithm would just build a list of all types referenced in includes/excludes, sort them (top classes first), then go through that list in order, applying filters as it goes:

* A is included => add all subclasses of A to `allIncluded`, remove them from `allExcluded`
* B is excluded => remove all subclasses of B from `allIncluded`, add them to `allExcluded`

Feel free to keep this algorithm though, I'm just mentioning this in case you have a hard time adapting it to my suggestion below.